### PR TITLE
Confluence Loader: Fix CQL loading

### LINKED ISSUE #12082: Confluence document loader not handling pagination correctly anymore (when using Confluence Cloud)
### System Info

LangChain 0.0.317
atlassian-python-api 3.41.3

### Who can help?

@hwchase17 

### Information

- [X] The official example notebooks/scripts
- [ ] My own modified scripts

### Related Components

- [ ] LLMs/Chat Models
- [ ] Embedding Models
- [ ] Prompts / Prompt Templates / Prompt Selectors
- [ ] Output Parsers
- [X] Document Loaders
- [ ] Vector Stores / Retrievers
- [ ] Memory
- [ ] Agents / Agent Executors
- [ ] Tools / Toolkits
- [ ] Chains
- [ ] Callbacks/Tracing
- [ ] Async

### Reproduction

When playing with documents from Confluence (Cloud hosting in my case), I have noticed that I am getting multiple copies of the same document in my vector store or from my Confluence retriever (custom implementation not in LangChain).  In both cases, the vector store is populated with LangChain's Confluence document loader and my retriever is also using that same document loader.  The problem happens in `ConfluenceLoader.paginate_request()` in the `while` loop until we reach `max_pages`.  In summary, you will get copies of the same documents until you reach `max_pages`.  For example, if your query (I'm using CQL in my use-case) returns 10 documents and `max_pages` is set to `200`, you will get 20 copies of each document.  This also makes the search process much slower.

The previous pagination system of Confluence REST server was leveraging an index controlled by the `start` parameter.  According to this [page](https://developer.atlassian.com/cloud/confluence/change-notice-moderize-search-rest-apis/), it has been deprecated in favor of a cursor system.  It is now recommended to follow `_links.next` instead of relying on the `start` parameter and getting an empty result when documents to be returned have been exhausted.  

This change is now in effect for the Cloud hosting of Confluence, not sure about private deployments.  You can see for yourself by running a script that looks like this:

```python
from atlassian import Confluence

site = "templates" # Public site 
confluence = Confluence(url=f"https://{site}.atlassian.net/")

cql='type=page AND text~"is"'
response_1 = confluence.cql(cql=cql, start=0)
print(response_1) # Easier to see in debug mode
next_start = len(response_1["results"])*1000
response_2 = confluence.cql(cql=cql, start=next_start)
print(response_2) # Notice returned documents are the same as in response_1
```

### Expected behavior

The Confluence document loader should handle properly pagination for both the old (using `start`) and new way (cursor) to maintain backward compatibility for private Confluence deployments.  It should also properly stop when `max_pages` has been reached.

### PR DESCRIPTION
**Pull request by Tonkonozhenko**
fix #12082

### CODE REVIEW

**Comment by Tonkonozhenko** [comment]
@ccurme could you review the PR, please?

**Comment by Tonkonozhenko** [comment]
@baskaryan, @efriis, @eyurtsev could you review the PR, please?

**Comment by Tonkonozhenko** [comment]
@efriis, @eyurtsev, @ccurme could you review the MR with the fix? Without it, it's completely impossible to load confluence by CQL

File: libs/community/langchain_community/document_loaders/confluence.py [hunk 1]
```diff
@@ -428,17 +428,25 @@ def lazy_load(self) -> Iterator[Document]:
         yield from self._lazy_load()
 
     def _search_content_by_cql(
-        self, cql: str, include_archived_spaces: Optional[bool] = None, **kwargs: Any
-    ) -> List[dict]:
-        url = "rest/api/content/search"
+        self,
+        cql: str,
+        include_archived_spaces: Optional[bool] = None,
+        nexturl: str = "",
```
[the remark this answers is not in the document]
Comment (Tonkonozhenko) [reply libs/community/langchain_community/document_loaders/confluence.py]: 👍🏼

File: libs/community/langchain_community/document_loaders/confluence.py [hunk 2]
```diff
@@ -428,17 +428,25 @@ def lazy_load(self) -> Iterator[Document]:
...
 
-        params: Dict[str, Any] = {"cql": cql}
-        params.update(kwargs)
-        if include_archived_spaces is not None:
-            params["includeArchivedSpaces"] = include_archived_spaces
+            response = self.confluence.get(url, params=params)
 
-        response = self.confluence.get(url, params=params)
-        return response.get("results", [])
+        return response.get("results", []), response["_links"].get("next", "")
```
[the remark this answers is not in the document]
Comment (Tonkonozhenko) [reply libs/community/langchain_community/document_loaders/confluence.py]: it shouldn't, but I've changed to .get

File: libs/community/langchain_community/document_loaders/confluence.py [hunk 3]
```diff
@@ -476,9 +485,15 @@ def paginate_request(self, retrieval_method: Callable, **kwargs: Any) -> List:
                 ),
                 before_sleep=before_sleep_log(logger, logging.WARNING),
             )(retrieval_method)
-            batch = get_pages(**kwargs, start=len(docs))
-            if not batch:
-                break
+            if self.cql:  # only cql
```
[the remark this answers is not in the document]
Comment (Tonkonozhenko) [reply libs/community/langchain_community/document_loaders/confluence.py]: Added the check for next_url presence. maybe some old versions of confluence don't have the cursor paginations

[the remark this answers is not in the document]
Comment (Tonkonozhenko) [reply libs/community/langchain_community/document_loaders/confluence.py]: It was a bad idea. Restored original version. Confluence cloud only tells us about cursor pagination.
https://developer.atlassian.com/cloud/confluence/rest/v1/api-group-content/#api-wiki-rest-api-content-search-get

self-hosted installation was deprecated on Feb 15, 2024
https://www.atlassian.com/migration/assess/journey-to-cloud

File: libs/community/langchain_community/document_loaders/confluence.py [hunk 4]
```diff
@@ -476,9 +485,15 @@ def paginate_request(self, retrieval_method: Callable, **kwargs: Any) -> List:
...
-            batch = get_pages(**kwargs, start=len(docs))
-            if not batch:
-                break
+            if self.cql:  # cursor pagination for CQL
+                batch, next_url = get_pages(**kwargs, next_url=next_url)
+                if not next_url:
+                    docs.extend(batch)
+                    break
+            else:
+                batch = get_pages(**kwargs, start=len(docs))
```
Comment (ccurme) [inline libs/community/langchain_community/document_loaders/confluence.py:494]: does this need to be `batch, _ = get(pages(...))`?

Comment (Tonkonozhenko) [reply libs/community/langchain_community/document_loaders/confluence.py:494]: no, because for other than cql, it doesn't return anything except batch

**Review by ccurme — APPROVED** [review]

**Pull request merged**